### ISSUE TITLE
[P2] Concept bottleneck + tiny causal LM: 明示概念を経由する最小生成モデルを作る

### ISSUE THREAD

**Issue by kazuaki-nakamura** [kazuaki-nakamura/norishio-lm#3]
## 目的
multi-channel encoderの出力を、直接next-token headへ流すのではなく、解釈可能なconcept bottleneckを明示的に経由してからsurface generationへ戻す最小モデルを実装する。

前提: Issue #1 と #2 が完了していること。未完了なら本Issueは設計のみ。

## 中核仮説
同じ意味を異なる表現で書いたとき、surface tokenが違ってもbottleneck表現が部分的に共有されるなら、paraphrase・compositional generalization・rare compound理解に利点が出る可能性がある。

最初の観測例:
- `また会いたい`
- `明日もいてほしい`
- `離れたくない`

これらを最初から同一意味と決め打ちしない。共通要素と差分（時間、願望、否定、分離回避など）を表現可能なschemaを設計する。

## 実装要件
1. 明示的なconcept schemaを定義する。例: entity / event / attribute / relation / intention / polarity / temporal等。固定語彙は小さくてよいが、コード内の謎のindex列にしない。
2. `ConceptBottleneck`を実装し、encoder stateからconcept logitsまたは構造化latentを生成する。
3. tiny causal LMまたは小型decoderを追加し、surface token lossを計算できるようにする。
4. auxiliary lossを個別にON/OFF可能にする。最低限 `L_lm`, `L_sense`, `L_sememe`, `L_concept`。未ラベル部分は適切にmaskする。
5. concept bypass baselineを必ず用意し、bottleneckを通さない経路との比較ができるようにする。
6. concept supervisionを完全教師ありに固定しない。ラベル欠損でもLM lossのみでforwardできる。
7. concept stateをinspect可能なAPIを持たせる。研究者が「何を経由したか」を確認できること。

## 最小実験
CPUで終わるtoy datasetを作り、以下を比較する:
- A: tiny LM baseline
- B: multi-channel encoder + LM, concept bottleneckなし
- C: multi-channel encoder + explicit concept bottleneck + LM

評価は少なくとも:
- train loss / held-out loss
- toy paraphrase clusteringまたはconcept-label accuracy
- misleading glyph cueへの耐性を確認する小テスト

小さいtoy実験から一般性能を主張しない。目的は配線と測定可能性の確認。

## 完了条件
- [ ] tiny CPU trainingが再現できるseed付きscriptとして存在する。
- [ ] baseline / bypass / bottleneckの3経路を同一ハーネスで比較できる。
- [ ] bottleneckのconcept出力を人間が読める形でdumpできる。
- [ ] loss componentごとの値をログでき、各lossをablationできる。
- [ ] 少なくとも1つのtoy held-out splitで結果を保存し、`docs/handoff.md`に数値と限界を書く。
- [ ] テストは外部ネットワークやGPUを要求しない。
- [ ] READMEに「toy modelでありLLM性能を示すものではない」と明記する。

## 非目標
- 7B等の大規模事前学習
- SOTA主張
- 人手で与えたconcept labelをモデル自身の発見と呼ぶこと
- 字源をsemantic ground truthとして使うこと

ブランチ案: `codex/concept-bottleneck-tiny-lm`。Issue #2のPRがmergeされた後に着手する。

---

## 着手指示・のり塩作成データの引き継ぎ（2026-09-10）

ユーザーが次段階への着手と、のり塩による教材作成を承認しました。PR #5はmasterへマージ済み、Issue #2はcompletedでクローズ済みです。以下は元の要件に対する具体化です。

### 提供データ

- データ準備ブランチ: `norishio/issue3-data-v1`
- 固定コミット: `e080504082946dbcc584b10f638eb066785db9b1`
- 基点: `bf0a0b94f8026b68f37b1f53833085e8ecdc04f5`（PR #5 merge）
- [教材説明](https://github.com/kazuaki-nakamura/norishio-lm/blob/e080504082946dbcc584b10f638eb066785db9b1/data/issue3/README.md)
- 原稿: `data/issue3/seed.json`
- 生成器・入力境界: `data/issue3/toy_corpus.py`
- 厳密ボトルネック用decoder入力: `data/issue3/strict_decoder.py`
- データ検査: `tests/test_issue3_data.py`、`tests/test_issue3_strict_decoder.py`
- 再生成照合: `data/issue3/expected-manifest.json`

最新masterから隔離したIssue #3作業ブランチを作り、上のデータコミットをcherry-pickして統合してください。worker既定の`codex/issue-3`等を使って構いません。既存作業があれば先に確認し、重複PRを作らないこと。データブランチをmasterへ自動mergeする必要はありません。モデル実装と合わせた1本のPRでレビューします。

データは新規AI作成のauthored_demo。会話履歴・Project添付資料・個人情報・外部辞書の取り込みではありません。train=450、validation=150、test=150、diagnostic=24。人物語×時点の組合せをgroupで分け、同一状況の全言い換え・対比を同じsplitに保持します。同じテンプレート群での未見組合せ評価であり、自然対話や未見テンプレートへの一般化を示すデータではありません。

```bash
python data/issue3/toy_corpus.py --check
python data/issue3/toy_corpus.py --out codex/work_output/norishio-toy-v1
python -m pytest -q -p no:cacheprovider tests/test_issue3_data.py tests/test_issue3_strict_decoder.py
```

出力先が既存なら別の出力先を選ぶか、manifestを検証して既存出力を利用する。無断で削除・上書きしない。生成JSONLはignoredのwork_outputへ置き、原稿とhashから再生成可能な形を保つ。

提供側の検証: Linux / Python 3.13.5 / pytest 9.0.2で上記新規2テストファイルが合計 **22 passed**。生成・分割・入力境界・teacher-forcing整列等を検査。GitHubに保存した7ファイルのblob SHAはローカル検査対象と照合済み。既存repo全体のテストやモデル学習は提供側では未実行。サブエージェントへの委譲なし。Codex側で統合後の全検証を実行してください。

### 入力漏洩と比較の必須検査

1. 最初は、ソース全文を観測してから参照文を出す**条件付き生成**。A/B/Cに同じ観測ソースと分割を与える。無条件LMの性能とは呼ばない。
2. `inputs.context/text`だけをencoder側へ渡す。`targets`、sample ID、group ID、split、作者情報、diagnosticの正解・interpretationを入力特徴にしない。辞書・feature vocabはtrainだけで構築し、参照文から補助情報を引かない。
3. **重要: strict Cのdecoderへソーストークンやソースのattention KVを渡さない。** `toy_corpus.teacher_forcing()`のソース付き系列はA/B用。Cのdecoderには`strict_decoder.target_history()`による過去の出力トークンと、encoderから**予測した**conceptだけを渡す。gold conceptやsurface latentを渡す経路はoracle/bypassとして別実験にする。
4. Cはconceptを固定した状態でソースの表記やencoder stateを差し替えてもdecoder出力が変わらないことを検査する。さらにconceptへの介入で出力がどう変わるかを記録する。hard/soft conceptの扱いを明記し、soft scoresが余分な情報を運ぶ可能性を検証限界に残す。
5. teacher forcingのラベルは1トークン先。未来targetを変えてもそれ以前のlogitsが変わらないこと。source-prefixとPADは適切にlossから除外する。源文全体の概念教師はソース末尾でのみ使い、途中のprefixへ先読みラベルを配らない。
6. `null`の補助注釈は欠損としてmaskする。全aux欠損でもLM lossだけでforward/backward可能にする。
7. datasetのNOT(WANT(...))とWANT(NOT(...))、予定と願望、反復明示、行為主体の差を潰さない。diagnostic 24件は本指標と分け、未学習カテゴリを固定headの正解IDとして無理に採点しない。誤字形fixtureを実際の字源として辞書へ混ぜない。
8. seed=7でCPU上の小さい再現実験から始め、A/B/Cの損失、対象数、パラメータ数、step数、実行時間、各loss/チャネルのablationを記録する。Cが負けてもデータや採点規則を都合よく変更しない。testは選択・調整に使わない。

### 完了・制約

AGENTS.mdと関連OKFを読み、全pytest、既存両デモ、新規学習デモ、必要なknowledge/OKF/MCP検査を実行。docs/handoff.mdに実際の数値・コマンド・SHA・限界を記録する。データの機械検査成功を、意味理解や学習効果の実証と混同しない。

有料GPU、大規模学習、外部辞書の取得、秘密情報や私的ログの追加、自動merge、Issueの自動closeはしない。作業報告とPRまでを今回の範囲とします。

**Comment by kazuaki-nakamura** [comment]
実装と検証を完了し、レビュー用PR #6を作成しました: https://github.com/kazuaki-nakamura/norishio-lm/pull/6

ブランチ `codex/issue-3`、head `d46abe8c8fec73fc495612425226d839727e7a41`。指定教材コミットを統合し、source-only入力、7項目の明示concept、strict C decoder、masked個別損失、A/B/C比較と介入を実装しました。

親の統合検証: 179 passed, 2 subtests passed（skipなし）、既存両デモ成功、教材manifest内容と4 JSONL SHA256一致、39原本のknowledge full/OKF/MCP成功。CPU seed=7/60stepの8条件を2回実行し、時間以外の全指標・trace・concept dumpが一致。test LMはA 2.191837、B 2.136616、C 2.087212。C concept項目accuracyは0.551429で、一般性能の主張ではありません。固定concept時のsource差替えと字形層除外時の誤字形差は0。

実際にgpt-5.6-lunaへモデル/損失と入力adapter/独立レビューを委譲。親がstrict Cのauxiliary混入を修正指示し、ハーネス・統合回帰・最終検査を担当しました。結果・コマンド・限界はdocs/handoff.mdに記録。単一seed、soft概念、未使用semantic入力層、自由生成品質は課題です。PRは未マージ、Issueはopenのai-review。定期チェック再開は行っていません。
